File: content_tool/models/persona.py
import re
from typing import Any, Literal

from pydantic import BaseModel, Field

# CJK detection (CJK Ext A + Unified Ideographs + Compatibility Ideographs).
# Mirrors the web heuristic so the persona-block label set is auto-derived from
# the voice's output language — no separate manual control.
_CJK_RE = re.compile(r"[㐀-鿿豈-﫿]")

GlossaryStatus = Literal["preferred", "avoid", "forbidden", "do_not_translate"]


class VoiceLocale(BaseModel):
    """Per-voice locale / brand identity, stored in ``personas.locale`` (JSONB).

    The defaults reproduce ``bowtie-editor`` (Bowtie HK 繁體中文) **byte-for-byte**,
    so an empty ``{}`` locale is a no-op — existing voices behave exactly as
    before. A voice opts into a different brand/language/market purely as data.

    Fields:
      - ``output_language`` — value for the ``{output_language}`` prompt token.
      - ``brand_name``      — value for the ``{brand_name}`` prompt token.
      - ``market``          — value for the ``{market}`` token (``topic_hot`` etc.).
      - ``sources_heading`` — explicit sources ``<h2>`` text; ``None`` keeps
        today's Traditional↔Simplified script auto-detection (safe for zh voices).
      - ``faq_heading``     — FAQ heading used by ``render_html``'s fallback/split.

    The persona-block label set (Traditional-Chinese vs English scaffolding) is
    **auto-derived** from ``output_language`` — see ``_labels_for``. There is no
    separate manual control.
    """

    output_language: str = "香港繁體中文"
    brand_name: str = "Bowtie"
    market: str = "Google 香港繁中"
    sources_heading: str | None = None
    faq_heading: str = "常見問題"

    @classmethod
    def from_raw(cls, raw: dict[str, Any] | None) -> "VoiceLocale":
        """Build from a raw JSONB value. ``None``/``{}`` → all HK-ZH defaults."""
        if not raw:
            return cls()
        return cls.model_validate(raw)


class PersonaBlockLabels(BaseModel):
    """Scaffolding labels for ``PersonaPack.to_prompt_block``.

    A label set auto-derived from ``VoiceLocale.output_language`` so a non-Chinese
    voice does not emit Traditional-Chinese scaffolding around its content. The
    ``zh-Hant`` set is byte-identical to the strings hardcoded before this change
    so HK-ZH voices are a no-op.
    """

    persona_header: str
    role: str
    voice_rules: str
    banned_terms: str
    required_phrasings: str
    tone_examples: str
    tone_good: str
    tone_bad: str
    glossary_header: str
    forbidden: str
    avoid: str
    avoid_arrow: str  # opener between an avoided term and its replacement target
    avoid_arrow_close: str  # closer after the replacement target
    avoid_no_target: str  # placeholder when no preferred target exists
    do_not_translate: str
    preferred_open: str  # opener before the preferred/term
    preferred_close: str  # closer after the preferred/term
    variants_open: str  # opener wrapping the variant list
    variants_close: str  # closer after the variant list


# zh-Hant (default) — byte-identical to the strings used before parameterization.
# NOTE: the required-phrasings label intentionally keeps the exact pre-change
# Traditional-Chinese bytes so the assembled HK-ZH prompt is unchanged; the
# brand/locale-neutral wording lives only in the English set below.
_LABELS_ZH_HANT = PersonaBlockLabels(
    persona_header="# 撰稿人格",
    role="角色：",  # noqa: RUF001
    voice_rules="語氣規則：",  # noqa: RUF001
    banned_terms="避免使用的字詞：",  # noqa: RUF001
    required_phrasings="必須採用的香港用語：",  # noqa: RUF001
    tone_examples="語氣示例：",  # noqa: RUF001
    tone_good="好：",  # noqa: RUF001
    tone_bad="壞：",  # noqa: RUF001
    glossary_header="# 詞彙表 · Glossary",
    forbidden="禁用：",  # noqa: RUF001
    avoid="避用：",  # noqa: RUF001
    avoid_arrow=" → 改用「",
    avoid_arrow_close="」",
    avoid_no_target="(無替代詞)",
    do_not_translate="保留原文：",  # noqa: RUF001
    preferred_open="用「",
    preferred_close="」",
    variants_open="（避用：",  # noqa: RUF001
    variants_close="）",  # noqa: RUF001
)

# en — English scaffolding; emits NO Traditional-Chinese labels.
_LABELS_EN = PersonaBlockLabels(
    persona_header="# Persona",
    role="Role: ",
    voice_rules="Voice rules:",
    banned_terms="Terms to avoid: ",
    required_phrasings="Required phrasings: ",
    tone_examples="Tone examples:",
    tone_good="Good: ",
    tone_bad="Bad: ",
    glossary_header="# Glossary",
    forbidden="Forbidden: ",
    avoid="Avoid: ",
    avoid_arrow=" → use \"",
    avoid_arrow_close="\"",
    avoid_no_target="(no alternative)",
    do_not_translate="Do not translate: ",
    preferred_open="Use \"",
    preferred_close="\"",
    variants_open=" (avoid: ",
    variants_close=")",
)


def _labels_for(output_language: str) -> PersonaBlockLabels:
    """Pick the persona-block label set from the voice's ``output_language``.

    Auto-derived: a Chinese output language (any CJK ideograph present) → the
    byte-identical Traditional-Chinese set; a non-Chinese (Latin-script) output
    language → English labels. This reproduces the previous explicit ``ui_lang``
    choice for every existing voice without a manual control.
    """
    return _LABELS_ZH_HANT if _CJK_RE.search(output_language) else _LABELS_EN


class GlossaryEntry(BaseModel):
    term: str
    preferred: str = ""
    variants: list[str] = Field(default_factory=list)
    status: GlossaryStatus = "preferred"
    notes: str | None = None

    def lookup_strings(self) -> list[str]:
        out = [self.term, *self.variants]
        if self.preferred:
            out.append(self.preferred)
        return [s for s in out if s]


class DisclaimerTemplate(BaseModel):
    condition: str = ""
    disclaimer: str = ""


class PersonaPack(BaseModel):
    name: str
    voice_rules: list[str]
    banned_terms: list[str]
    required_phrasings: list[str]
    disclaimer_templates: dict[str, DisclaimerTemplate]
    tone_examples: dict[str, list[str]]
    glossary: list[GlossaryEntry] = Field(default_factory=list)
    locale: VoiceLocale = Field(default_factory=VoiceLocale)

    def to_prompt_block(self, context_text: str | None = None) -> str:
        """Render as a Chinese-language persona block for system prompts.

        When ``context_text`` is supplied the glossary section is filtered to
        only entries whose term/variants/preferred form substring-match the
        context. Keeps prompts bounded for large termbases while still
        surfacing the entries that matter for the current brief/draft.
        """
        lbl = _labels_for(self.locale.output_language)
        good = "\n".join(f"  {lbl.tone_good}{x}" for x in self.tone_examples.get("good", []))
        bad = "\n".join(f"  {lbl.tone_bad}{x}" for x in self.tone_examples.get("bad", []))
        glossary_section = self._render_glossary(context_text, lbl)
        return (
            f"{lbl.persona_header}\n"
            f"{lbl.role}{self.name}\n"
            f"{lbl.voice_rules}\n" + "\n".join(f"- {r}" for r in self.voice_rules) + "\n"
            f"{lbl.banned_terms}{', '.join(self.banned_terms)}\n"
            f"{lbl.required_phrasings}{', '.join(self.required_phrasings)}\n"
            f"{lbl.tone_examples}\n{good}\n{bad}\n"
            f"{glossary_section}"
        )

    def _render_glossary(
        self, context_text: str | None, lbl: PersonaBlockLabels | None = None
    ) -> str:
        if lbl is None:
            lbl = _labels_for(self.locale.output_language)
        entries = self._filter_glossary(context_text)
        if not entries:
            return ""
        lines: list[str] = [lbl.glossary_header]
        for e in entries:
            variants = (
                f"{lbl.variants_open}{', '.join(e.variants)}{lbl.variants_close}"
                if e.variants
                else ""
            )
            note = f" — {e.notes}" if e.notes else ""
            if e.status == "forbidden":
                lines.append(f"- {lbl.forbidden}{e.term}{variants}{note}")
            elif e.status == "avoid":
                target = e.preferred or lbl.avoid_no_target
                lines.append(
                    f"- {lbl.avoid}{e.term}{lbl.avoid_arrow}{target}"
                    f"{lbl.avoid_arrow_close}{variants}{note}"
                )
            elif e.status == "do_not_translate":
                lines.append(f"- {lbl.do_not_translate}{e.term}{note}")
            else:
                target = e.preferred or e.term
                lines.append(
                    f"- {lbl.preferred_open}{target}{lbl.preferred_close}{variants}{note}"
                )
        return "\n".join(lines) + "\n"

    def _filter_glossary(self, context_text: str | None) -> list[GlossaryEntry]:
        if context_text is None:
            return list(self.glossary)
        haystack = context_text.lower()
        return [
            e for e in self.glossary
            if any(s.lower() in haystack for s in e.lookup_strings())
        ]

File: deploy/cloudflare-workers/src/agents/persona.ts
// Persona prompt-block rendering — TypeScript port of:
//   content_tool/models/persona.py  (PersonaPack, GlossaryEntry, to_prompt_block)
//   content_tool/policy/personas.py (load_persona)
//
// Template strings are reproduced byte-for-byte from the Python source so that
// the prompts sent to Gemini are identical across runtimes.

import type { Sql } from "postgres";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type GlossaryStatus = "preferred" | "avoid" | "forbidden" | "do_not_translate";

export interface GlossaryEntry {
  term: string;
  preferred: string;
  variants: string[];
  status: GlossaryStatus;
  notes: string | null;
}

export interface DisclaimerTemplate {
  condition: string;
  disclaimer: string;
}

/**
 * Per-voice locale / brand identity — TS mirror of
 * `content_tool.models.persona.VoiceLocale`. Stored in `personas.locale`
 * (JSONB). The defaults reproduce `bowtie-editor` (Bowtie HK 繁體中文)
 * byte-for-byte, so an empty `{}` locale is a no-op.
 */
export interface VoiceLocale {
  outputLanguage: string;
  brandName: string;
  market: string;
  sourcesHeading: string | null;
  faqHeading: string;
}

/** HK-ZH defaults (mirror VoiceLocale's Python field defaults). */
export function defaultVoiceLocale(): VoiceLocale {
  return {
    outputLanguage: "香港繁體中文",
    brandName: "Bowtie",
    market: "Google 香港繁中",
    sourcesHeading: null,
    faqHeading: "常見問題",
  };
}

interface RawVoiceLocale {
  output_language?: string;
  brand_name?: string;
  market?: string;
  sources_heading?: string | null;
  faq_heading?: string;
}

/** Build a VoiceLocale from a raw JSONB value. null/{}/non-object → defaults. */
export function voiceLocaleFromRaw(raw: unknown): VoiceLocale {
  const d = defaultVoiceLocale();
  if (raw === null || typeof raw !== "object") {
    return d;
  }
  const r = raw as RawVoiceLocale;
  return {
    outputLanguage: r.output_language ?? d.outputLanguage,
    brandName: r.brand_name ?? d.brandName,
    market: r.market ?? d.market,
    sourcesHeading: r.sources_heading ?? d.sourcesHeading,
    faqHeading: r.faq_heading ?? d.faqHeading,
  };
}

/** Raw snake_case shape of a VoiceLocale (wire/JSONB form). */
export interface RawVoiceLocaleOut {
  output_language: string;
  brand_name: string;
  market: string;
  sources_heading: string | null;
  faq_heading: string;
}

/** Serialize a VoiceLocale back to its snake_case wire shape (camel → snake). */
export function voiceLocaleToRaw(loc: VoiceLocale): RawVoiceLocaleOut {
  return {
    output_language: loc.outputLanguage,
    brand_name: loc.brandName,
    market: loc.market,
    sources_heading: loc.sourcesHeading,
    faq_heading: loc.faqHeading,
  };
}

export interface PersonaPack {
  name: string;
  voiceRules: string[];
  bannedTerms: string[];
  requiredPhrasings: string[];
  disclaimerTemplates: Record<string, DisclaimerTemplate>;
  toneExamples: { good: string[]; bad: string[] };
  glossary: GlossaryEntry[];
  locale: VoiceLocale;
}

/**
 * Unsaved persona-draft overrides for the editor preview. Each field is
 * optional and absent ⇒ the persona's stored value (byte-identical to today).
 * Folds the previously-standalone `localeOverride` together with the new
 * `glossary` draft into one object threaded through `defaultPersonaBlock` /
 * `toPromptBlock`.
 */
export interface PersonaOverride {
  locale?: VoiceLocale;
  glossary?: GlossaryEntry[];
}

// ---------------------------------------------------------------------------
// Persona-block scaffolding labels (auto-derived from VoiceLocale.outputLanguage)
// ---------------------------------------------------------------------------

/**
 * Scaffolding labels for `toPromptBlock` — TS mirror of Python
 * `PersonaBlockLabels`. The `zh-Hant` set is byte-identical to the strings used
 * before parameterization so HK-ZH voices are a no-op; the `en` set emits no
 * Traditional-Chinese scaffolding.
 */
interface PersonaBlockLabels {
  personaHeader: string;
  role: string;
  voiceRules: string;
  bannedTerms: string;
  requiredPhrasings: string;
  toneExamples: string;
  toneGood: string;
  toneBad: string;
  glossaryHeader: string;
  forbidden: string;
  avoid: string;
  avoidArrow: string; // between term and target
  avoidArrowClose: string; // after target
  avoidNoTarget: string; // placeholder when no preferred target exists
  doNotTranslate: string;
  preferredOpen: string; // before preferred/term
  preferredClose: string; // after preferred/term
  variantsOpen: string; // wraps variant list (open)
  variantsClose: string; // wraps variant list (close)
}

// zh-Hant (default) — byte-identical to the pre-change hardcoded strings.
// NOTE: requiredPhrasings keeps the exact bytes "必須採用的香港用語：" so the
// assembled HK-ZH prompt is unchanged; the neutral wording lives only in `en`.
const LABELS_ZH_HANT: PersonaBlockLabels = {
  personaHeader: "# 撰稿人格",
  role: "角色：",
  voiceRules: "語氣規則：",
  bannedTerms: "避免使用的字詞：",
  requiredPhrasings: "必須採用的香港用語：",
  toneExamples: "語氣示例：",
  toneGood: "好：",
  toneBad: "壞：",
  glossaryHeader: "# 詞彙表 · Glossary",
  forbidden: "禁用：",
  avoid: "避用：",
  avoidArrow: " → 改用「",
  avoidArrowClose: "」",
  avoidNoTarget: "(無替代詞)",
  doNotTranslate: "保留原文：",
  preferredOpen: "用「",
  preferredClose: "」",
  variantsOpen: "（避用：",
  variantsClose: "）",
};

// en — English scaffolding; emits NO Traditional-Chinese labels.
const LABELS_EN: PersonaBlockLabels = {
  personaHeader: "# Persona",
  role: "Role: ",
  voiceRules: "Voice rules:",
  bannedTerms: "Terms to avoid: ",
  requiredPhrasings: "Required phrasings: ",
  toneExamples: "Tone examples:",
  toneGood: "Good: ",
  toneBad: "Bad: ",
  glossaryHeader: "# Glossary",
  forbidden: "Forbidden: ",
  avoid: "Avoid: ",
  avoidArrow: ' → use "',
  avoidArrowClose: '"',
  avoidNoTarget: "(no alternative)",
  doNotTranslate: "Do not translate: ",
  preferredOpen: 'Use "',
  preferredClose: '"',
  variantsOpen: " (avoid: ",
  variantsClose: ")",
};

// CJK detection (CJK Ext A + Unified + Compatibility Ideographs) — mirrors the
// Python `_CJK_RE` so both backends auto-derive the label set identically.
const CJK_RE = /[㐀-鿿豈-﫿]/;

/**
 * Pick the persona-block label set from the voice's `outputLanguage`
 * (mirror of Python `_labels_for`). Auto-derived: a Chinese output language
 * (any CJK ideograph) → Traditional-Chinese labels; a non-Chinese (Latin-script)
 * output language → English labels. No separate manual control.
 */
function labelsFor(outputLanguage: string): PersonaBlockLabels {
  return CJK_RE.test(outputLanguage) ? LABELS_ZH_HANT : LABELS_EN;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/** Strings used for context-text substring matching (mirrors lookup_strings()). */
function lookupStrings(entry: GlossaryEntry): string[] {
  const candidates: string[] = [entry.term, ...entry.variants];
  if (entry.preferred) {
    candidates.push(entry.preferred);
  }
  return candidates.filter((s) => s.length > 0);
}

/** Filter glossary to entries that substring-match contextText (case-insensitive). */
function filterGlossary(glossary: GlossaryEntry[], contextText: string | undefined): GlossaryEntry[] {
  if (contextText === undefined) {
    return [...glossary];
  }
  const haystack = contextText.toLowerCase();
  return glossary.filter((e) =>
    lookupStrings(e).some((s) => haystack.includes(s.toLowerCase())),
  );
}

/** Render glossary section — mirrors PersonaPack._render_glossary(). */
function renderGlossary(
  glossary: GlossaryEntry[],
  contextText: string | undefined,
  lbl: PersonaBlockLabels,
): string {
  const entries = filterGlossary(glossary, contextText);
  if (entries.length === 0) {
    return "";
  }
  const lines: string[] = [lbl.glossaryHeader];
  for (const e of entries) {
    const variants =
      e.variants.length > 0 ? `${lbl.variantsOpen}${e.variants.join(", ")}${lbl.variantsClose}` : "";
    const note = e.notes ? ` — ${e.notes}` : "";
    if (e.status === "forbidden") {
      lines.push(`- ${lbl.forbidden}${e.term}${variants}${note}`);
    } else if (e.status === "avoid") {
      const target = e.preferred || lbl.avoidNoTarget;
      lines.push(`- ${lbl.avoid}${e.term}${lbl.avoidArrow}${target}${lbl.avoidArrowClose}${variants}${note}`);
    } else if (e.status === "do_not_translate") {
      lines.push(`- ${lbl.doNotTranslate}${e.term}${note}`);
    } else {
      // "preferred" (default)
      lines.push(`- ${lbl.preferredOpen}${e.preferred || e.term}${lbl.preferredClose}${variants}${note}`);
    }
  }
  return lines.join("\n") + "\n";
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Render the persona as a Chinese-language system-prompt block.
 *
 * When `contextText` is supplied the glossary section is filtered to only
 * entries whose term/variants/preferred form substring-match the context.
 * This mirrors Python's `PersonaPack.to_prompt_block(context_text)` exactly.
 */
export function toPromptBlock(persona: PersonaPack, contextText?: string): string {
  const lbl = labelsFor(persona.locale.outputLanguage);
  const good = persona.toneExamples.good.map((x) => `  ${lbl.toneGood}${x}`).join("\n");
  const bad = persona.toneExamples.bad.map((x) => `  ${lbl.toneBad}${x}`).join("\n");
  const glossarySection = renderGlossary(persona.glossary, contextText, lbl);

  return (
    `${lbl.personaHeader}\n` +
    `${lbl.role}${persona.name}\n` +
    `${lbl.voiceRules}\n` +
    persona.voiceRules.map((r) => `- ${r}`).join("\n") +
    `\n` +
    `${lbl.bannedTerms}${persona.bannedTerms.join(", ")}\n` +
    `${lbl.requiredPhrasings}${persona.requiredPhrasings.join(", ")}\n` +
    `${lbl.toneExamples}\n${good}\n${bad}\n` +
    glossarySection
  );
}

// ---------------------------------------------------------------------------
// DB row shape (jsonb columns arrive as parsed objects from postgres.js)
// ---------------------------------------------------------------------------

interface RawDisclaimerTemplate {
  condition?: string;
  disclaimer?: string;
}

interface RawGlossaryEntry {
  term: string;
  preferred?: string;
  variants?: string[];
  status?: GlossaryStatus;
  notes?: string | null;
}

/**
 * Map a raw (snake_case / JSONB) glossary value to validated `GlossaryEntry[]`.
 * Shared by `rowToPack` (DB row) and the preview's `glossary` draft override
 * (request body). A non-array value ⇒ `[]`; each entry is defaulted exactly as
 * the persona row mapping does, so a draft and a stored glossary render the
 * same `toPromptBlock` bytes.
 */
export function glossaryFromRaw(raw: unknown): GlossaryEntry[] {
  if (!Array.isArray(raw)) return [];
  // `term` MUST be a string — untrusted preview JSON can carry non-string fields
  // (route → 422 caps read `.length`), so narrow defensively and coerce the rest.
  return (raw as unknown[])
    .filter(
      (e): e is RawGlossaryEntry =>
        e !== null && typeof e === "object" && typeof (e as RawGlossaryEntry).term === "string",
    )
    .map((e) => ({
      term: e.term,
      preferred: typeof e.preferred === "string" ? e.preferred : "",
      variants: Array.isArray(e.variants) ? e.variants.filter((v) => typeof v === "string") : [],
      status: e.status ?? "preferred",
      notes: typeof e.notes === "string" ? e.notes : null,
    }));
}

interface RawToneExamples {
  good?: string[];
  bad?: string[];
}

/** Map a raw DB row to a validated PersonaPack. */
function rowToPack(row: {
  name: string;
  voice_rules: unknown;
  banned_terms: unknown;
  required_phrasings: unknown;
  disclaimer_templates: unknown;
  tone_examples: unknown;
  glossary: unknown;
  locale: unknown;
}): PersonaPack {
  const voiceRules = row.voice_rules as string[];
  const bannedTerms = row.banned_terms as string[];
  const requiredPhrasings = row.required_phrasings as string[];
  const rawDisclaimers = row.disclaimer_templates as Record<string, RawDisclaimerTemplate>;
  const rawTone = row.tone_examples as RawToneExamples;

  const disclaimerTemplates: Record<string, DisclaimerTemplate> = {};
  for (const [key, val] of Object.entries(rawDisclaimers)) {
    disclaimerTemplates[key] = {
      condition: val.condition ?? "",
      disclaimer: val.disclaimer ?? "",
    };
  }

  const glossary: GlossaryEntry[] = glossaryFromRaw(row.glossary ?? []);

  return {
    name: row.name,
    voiceRules,
    bannedTerms,
    requiredPhrasings,
    disclaimerTemplates,
    toneExamples: {
      good: rawTone.good ?? [],
      bad: rawTone.bad ?? [],
    },
    glossary,
    locale: voiceLocaleFromRaw(row.locale),
  };
}

/**
 * Load a persona from `content_tool.personas` by slug.
 *
 * Throws if no row is found (mirrors Python's scalar_one_or_none + YAML
 * fallback path — but in the Workers runtime there is no filesystem fallback,
 * so a missing slug is an unrecoverable error).
 */
export async function loadPersona(
  sql: Sql,
  slug: string,
): Promise<PersonaPack> {
  const rows = await sql<
    Array<{
      name: string;
      voice_rules: unknown;
      banned_terms: unknown;
      required_phrasings: unknown;
      disclaimer_templates: unknown;
      tone_examples: unknown;
      glossary: unknown;
      locale: unknown;
    }>
  >`
    SELECT name, voice_rules, banned_terms, required_phrasings,
           disclaimer_templates, tone_examples, glossary, locale
    FROM content_tool.personas
    WHERE slug = ${slug}
      AND is_archived = false
    LIMIT 1
  `;

  const row = rows[0];
  if (row === undefined) {
    throw new Error(`Persona not found: ${slug}`);
  }

  return rowToPack(row);
}
